stdlib/camlinternalMod.ml

```diff
@@ -51,8 +51,12 @@ let rec init_mod loc shape =
 let rec update_mod shape o n =
   match shape with
   | Function ->
-      if Obj.tag n = Obj.closure_tag && Obj.size n <= Obj.size o
-      then begin overwrite o n; Obj.truncate o (Obj.size n) (* PR#4008 *) end
+      (* The optimisation below is invalid on bytecode since
+         the RESTART instruction checks the length of closures.
+         See PR#4008 *)
+      if Sys.backend_type = Sys.Native &&
+        Obj.tag n = Obj.closure_tag && Obj.size n <= Obj.size o
```

Comment:
Apologies for splitting hair on style and indentation, but I'd rather see
```
if Sys.backend_type = Sys.Native
&& Obj.tag_n = ...
then ...
```
or, at least,
```
if Sys.backend... &&
   Obj.tag n 
```
(3 spaces)

Reply:
OK. Changed to the first option.